drivers/display/display_ili9xxx.c

```diff
@@ -141,14 +184,93 @@ static int ili9xxx_write(const struct device *dev, const uint16_t x,
 	return 0;
 }
 
+#ifdef CONFIG_ILI9XXX_READ
+
 static int ili9xxx_read(const struct device *dev, const uint16_t x,
 			const uint16_t y,
 			const struct display_buffer_descriptor *desc, void *buf)
 {
-	LOG_ERR("Reading not supported");
-	return -ENOTSUP;
+	const struct ili9xxx_config *config = dev->config;
+	struct ili9xxx_data *data = dev->data;
+	struct display_buffer_descriptor mipi_desc;
+	int r;
+	uint32_t gram_data, read_cnt, nbr_of_reads;
+	uint16_t *read_data_start = (uint16_t *)buf;
+
+	if (data->pixel_format != PIXEL_FORMAT_RGB_565) {
+		/* Only RGB565 can be supported, see note below */
+		return -ENOTSUP;
+	}
+
+	__ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
+	__ASSERT((desc->pitch * data->bytes_per_pixel * desc->height) <=
+			 desc->buf_size,
+		 "Output buffer to small");
+
+	LOG_DBG("Reading %dx%d (w,h) @ %dx%d (x,y)", desc->width, desc->height,
+		x, y);
+
+	r = ili9xxx_set_mem_area(dev, x, y, desc->width, desc->height);
+	if (r < 0) {
+		return r;
+	}
+
+	/*
+	 * ILI9XXX stores all pixel data in graphics ram (GRAM) as 18 bit
+	 * values. When using RGB565 pixel format, pixels are converted to
+	 * 18 bit values via a lookup table. When using RGB888 format, the
+	 * lower 2 bits of each pixel are simply dropped. When reading pixels,
+	 * the response format will always look like so:
+	 * | R[5:0] | x | x | G[5:0] | x | x | B[5:0] | x | x |
+	 * Where x represents "don't care". The internal format of the
+	 * ILI9XXX graphics RAM results in the following restrictions:
+	 * - RGB888 mode can't be supported.
+	 * - we can only read one pixel at once (since we need to do
+	 *   byte manipulation on the output)
+	 */
+
+	/* Setup MIPI descriptor to read 3 bytes (one pixel in GRAM) */
+	mipi_desc.width = 1;
+	mipi_desc.height = 1;
+	/* Per MIPI API, pitch must always match width */
+	mipi_desc.pitch = 1;
+
+	nbr_of_reads = desc->width * desc->height;
+
+	/* Initial read command should consist of RAMRD command, plus
+	 * 8 dummy clock cycles
+	 */
+	uint8_t cmd[] = {ILI9XXX_RAMRD, 0xFF};
+
+	for (read_cnt = 0; read_cnt < nbr_of_reads; read_cnt++) {
+		r = mipi_dbi_command_read(config->mipi_dev,
```

Comment:
So far, all the MIPI DBI display controllers I have worked with, and in contrast to `mipi_dbi_command_write`, require that command, dummy clocks and response be bound to the same CS activation cycle. So, I was wondering if the driver implementation enforces this requirement, implicitly. `SPI_HOLD_ON_CS` can then be used if reading data/pixels requires multiple calls to `mipi_dbi_command_read`.

Reply:
I saw the same requirement here- it is why I implemented this support, to verify `mipi_dbi_command_read` worked. You can see that the `mipi_dbi_command_read` implementation sets `SPI_HOLD_ON_CS`: https://github.com/zephyrproject-rtos/zephyr/pull/64587/files#diff-0c6d6eb06df9d6b88a3147689182e0f296a6d95d7b8d3576bcefef8c504424d1R201

The fact that `SPI_HOLD_ON_CS` must be set is an unfortunate side effect of abstracting the SPI controller interface behind the MIPI DBI API. I suppose we could expose support for this feature via some kind of `MIPI_HOLD_ON_CS` bit, but to me it makes the most sense to instead implement it implicitly within the `mipi_dbi_command_read` function, since most controllers need the CS line held low for reads anyways

Reply:
Thanks for clarifying this. I totally agree with your approach.